pilot/pkg/ambient/controller/waypoint_controller.go

```diff
@@ -130,103 +143,85 @@ func (rc *WaypointProxyController) Reconcile(name types.NamespacedName) error {
-	serviceAccounts, _ := rc.serviceAccounts.ServiceAccounts(gw.Namespace).List(klabels.Everything())
+	serviceAccounts, _ := rc.serviceAccounts.ServiceAccounts(gw.Namespace).
+		List(klabels.Everything())
 	for _, sa := range serviceAccounts {
 		if gatewaySA != "" && sa.Name != gatewaySA {
-			log.Debugf("skip service account %v, doesn't match gateway %v", sa.Name, gatewaySA)
+			istiolog.Debugf("skip service account %v, doesn't match gateway %v", sa.Name, gatewaySA)
 			continue
 		}
 		wantProxies.Insert(sa.Name)
 	}
 
-	add, remove := wantProxies.Diff(haveProxies)
-
-	if len(remove)+len(add) == 0 {
-		log.Debugf("reconcile: remove %d, add %d. Have %d", len(remove), len(add), len(haveProxies))
-	} else {
-		log.Infof("reconcile: remove %d, add %d. Have %d", len(remove), len(add), len(haveProxies))
-	}
-	for _, k := range remove {
-		log.Infof("removing waypoint proxy %q", k+"-waypoint-proxy")
-		if err := rc.client.Kube().AppsV1().Deployments(gw.Namespace).Delete(context.Background(), k+"-waypoint-proxy", metav1.DeleteOptions{}); err != nil {
-			return fmt.Errorf("pod remove: %v", err)
-		}
-
-		msg := fmt.Sprintf("Removed waypoint proxy from %q namespace", gw.Namespace)
-		if gatewaySA != "" {
-			msg += fmt.Sprintf(" for %q service account", gatewaySA)
-		}
-		err := rc.UpdateStatus(gw, map[string]*istiogw.Condition{
-			string(v1alpha2.GatewayConditionAccepted): {
-				Reason:  string(v1alpha2.GatewayReasonAccepted),
-				Message: msg,
-			},
-		})
-		if err != nil {
-			log.Errorf("unable to update Gateway status %v on delete: %v", gw.Name, err)
-		}
-	}
-	for _, k := range add {
-		log.Infof("adding waypoint proxy %v", k+"-waypoint-proxy")
+	for _, sa := range wantProxies.UnsortedList() {
 		input := MergedInput{
 			Namespace:      gw.Namespace,
 			GatewayName:    gw.Name,
 			UID:            string(gw.UID),
-			ServiceAccount: k,
+			ServiceAccount: sa,
 			Cluster:        rc.cluster.String(),
 			ProxyConfig:    rc.injectConfig().MeshConfig.GetDefaultConfig(),
 		}
-		proxyDeploy, err := rc.RenderDeploymentMerged(input)
+		proxyDeploy, err := rc.RenderDeploymentApply(input)
 		if err != nil {
 			return err
 		}
-
-		if _, err := rc.client.Kube().AppsV1().Deployments(name.Namespace).Create(context.Background(), proxyDeploy, metav1.CreateOptions{}); err != nil {
-			return fmt.Errorf("pod create: %v", err)
-		}
-
-		msg := fmt.Sprintf("Deployed waypoint proxy to %q namespace", gw.Namespace)
-		if gatewaySA != "" {
-			msg += fmt.Sprintf(" for %q service account", gatewaySA)
-		}
-		err = rc.UpdateStatus(gw, map[string]*istiogw.Condition{
-			string(v1alpha2.GatewayConditionReady): {
-				Reason:  string(v1alpha2.GatewayReasonReady),
-				Message: msg,
-			},
-			string(v1alpha2.GatewayConditionAccepted): {
-				Reason:  string(v1alpha2.GatewayReasonAccepted),
-				Message: msg,
-			},
-		})
+		// TODO: should support HPA, PDB, maybe others...
+		istiolog.Infof("Updating waypoint proxy %q", sa+"-waypoint-proxy")
```

Comment:
I am also seeing this logged 2x for each Gateway change. Not sure if its a regression though

Reply:
Well we do have to put the full Deployment spec in the request so its a bit more expensive? I suppose it makes sense, curious if other controllers somehow handle this. I suppose some debounce would do it. Not a high priority